src/Transformer.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\fb_instant_articles\Transformer.
+ */
+
+namespace Drupal\fb_instant_articles;
+
+/**
+ * Encapsulates Drupal-specific logic when using the Transformer class.
+ *
+ * Class Transformer
+ * @package Drupal\fb_instant_articles
+ */
+class Transformer extends \Facebook\InstantArticles\Transformer\Transformer {
+
+  /**
+   * Transformer constructor.
+   *
+   * Adds Drupal hook to allow transformer rules to be added by other modules.
+   * We invoke the hook in a constructor - as opposed to inside transform() or a
+   * new invokeRules() method - because of the way Transformer is normally used.
+   *
+   * @see hook_fb_instant_articles_transformer_rules()
+   *
+   * Note the parent class does not have a constructor, so we do not call
+   * @code parent::__construct() @endcode.
+   */
+  public function __construct() {
+    $rules = module_invoke_all('fb_instant_articles_transformer_rules');
```

Comment:
Hum, been testing and something still isn't right.  If I implement the hook that is fired here: https://github.com/scottrigby/module-fb_instant_articles/blob/2703063_refactor_base_classes_for_extensibility/src/InstantArticle.php#L27, the `$this` parameter that is passed there is not populated with all pieces of the Instant Article that should have been added via the field formatters.

The reason is, we effectively have two instances of `\Facebook\InstantArticles\Elements\InstantArticle` floating around.  The first is a `\Facebook\InstantArticles\Elements\InstantArticle` proper and the second is a `Drupal\fb_instant_articles_display\InstantArticle`.  The `\Facebook\InstantArticles\Elements\InstantArticle` class is created here: https://github.com/scottrigby/module-fb_instant_articles/blob/2703063_refactor_base_classes_for_extensibility/modules/fb_instant_articles_display/src/InstantArticle.php#L83.  The second, is created here, and the first is stored as a member of the second: https://github.com/scottrigby/module-fb_instant_articles/blob/2703063_refactor_base_classes_for_extensibility/modules/fb_instant_articles_display/src/InstantArticle.php#L115.  

The field formatters then act on the `\Facebook\InstantArticles\Elements\InstantArticle` object instance, eg. https://github.com/scottrigby/module-fb_instant_articles/blob/2703063_refactor_base_classes_for_extensibility/modules/fb_instant_articles_display/src/InstantArticle.php#L138,L139.  But then the alter hook passes the `Drupal\fb_instant_articles_display\InstantArticle` object instance: https://github.com/scottrigby/module-fb_instant_articles/blob/2703063_refactor_base_classes_for_extensibility/src/InstantArticle.php#L27.

Hopefully that makes sense, it's confusing to keep straight b/c there are three classes named `InstantArticle` all under different namespaces.

Trying to think about a resolution.